metadata/src/utils/validation.rs

```diff
@@ -262,53 +300,95 @@ fn get_extrinsic_hash(
         bytes = concat_and_hash4(
             &bytes,
             &hash(signed_extension.identifier.as_bytes()),
-            &get_type_hash(registry, signed_extension.extra_ty, &mut cache),
-            &get_type_hash(registry, signed_extension.additional_ty, &mut cache),
+            &get_type_hash(registry, signed_extension.extra_ty, outer_enum_hashes),
+            &get_type_hash(registry, signed_extension.additional_ty, outer_enum_hashes),
         )
     }
 
     bytes
 }
 
-/// Obtain the hash representation of the `frame_metadata::v15::OuterEnums`.
-fn get_outer_enums_hash(
-    registry: &PortableRegistry,
-    enums: &OuterEnumsMetadata,
-    only_these_variants: Option<&[&str]>,
-) -> [u8; HASH_LEN] {
-    /// Hash the provided enum type.
-    fn get_enum_hash(
-        registry: &PortableRegistry,
-        id: u32,
-        only_these_variants: Option<&[&str]>,
-    ) -> [u8; HASH_LEN] {
-        let ty = registry
-            .types
-            .get(id as usize)
-            .expect("Metadata should contain enum type in registry");
-
-        if let TypeDef::Variant(variant) = &ty.ty.type_def {
-            get_type_def_variant_hash(registry, variant, only_these_variants, &mut HashMap::new())
-        } else {
-            get_type_hash(registry, id, &mut HashMap::new())
+/// Hash representations of the `frame_metadata::v15::OuterEnums`.
+pub struct OuterEnumHashes {
+    call_hash: (u32, Hash),
+    error_hash: (u32, Hash),
+    event_hash: (u32, Hash),
+}
+
+impl OuterEnumHashes {
```

Comment:
Small thing; perhaps move this to a separate `outer_enum_hashes.rs` file, just so that the "pub"s are respected :)

Looks great otherwise though!